File: CPP/ByRound/2090/2090_D_Simple_Permutation.cpp
// Problem: CF 2090 D - Simple Permutation
// https://codeforces.com/contest/2090/problem/D

/*
D. Simple Permutation
Purpose: Construct a permutation of length n such that at least floor(n/3) - 1 of the ceiling averages c_i are prime.
Algorithms/Techniques: 
    - Brute-force search for largest prime less than or equal to n/2
    - Use that prime as a base to generate a sequence where values are arranged in a specific pattern
    - Check primality of computed averages using a helper function
Time Complexity: O(n * sqrt(n)) per test case due to primality check inside loop
Space Complexity: O(1) excluding input/output

Input:
    Integer t (number of test cases)
    For each test case, integer n (size of permutation)
Output:
    Permutation satisfying the condition
*/

#include <stdio.h>
#include <iosfwd>

#define ll long long
using namespace std;

ll t, n;

// Checks if a number is prime
int isp(ll x) {
  if (x == 1)
    return 0;
  for (ll i = 2; i * i <= x; i++) {
    if (x % i == 0)
      return 0;
  }
  return 1;
}

int main() {
  scanf("%d", &t);
  for (int i = 1; i <= t; i++) {
    scanf("%d", &n);
    if (n == 2)
      printf("2 1\n"); // Special case for n=2
    else if (n == 3)
      printf("2 1 3\n"); // Special case for n=3
    else {
      for (int i = n / 2; i > 0; i--) {
        if (isp(i)) { // Find largest prime <= n/2
          printf("%d ", i); // Print the prime number
          for (int j = i - 1; j > 0; j--)
            printf("%d %d ", j, i - j + i); // Generate pattern based on prime value
          for (int j = i + i; j <= n; j++)
            printf("%d ", j); // Continue with remaining numbers
          printf("\n");
          break;
        }
      }
    }
  }
  return 0;
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/D/2090_D_Simple_Permutation.py
# Problem: CF 2090 D - Simple Permutation
# https://codeforces.com/contest/2090/problem/D

"""
D. Simple Permutation

Purpose:
This code constructs a permutation of integers from 1 to n such that among the ceiling averages c_i,
there are at least floor(n/3) - 1 prime numbers.

Algorithm:
- Precompute primes up to 100000 using Sieve of Eratosthenes.
- For a given n, find the largest prime number less than or equal to 2*n//3.
- Construct the result list by:
  1. Starting with that prime
  2. Expanding outwards from the prime (left and right)
  3. Adding remaining numbers in order

Time Complexity: O(n * log(log(n))) for sieve + O(n) for construction = O(n * log(log(n)))
Space Complexity: O(n) for storing sieve and result list.

"""

import math

limit = 100000


def soe(limit):
    # Sieve of Eratosthenes to find all primes up to limit
    prime = [True] * (limit + 1)
    prime[0] = prime[1] = False
    for i in range(2, int(math.sqrt(limit)) + 1):
        if prime[i]:
            for j in range(i * i, limit + 1, i):  # Mark multiples of prime as not prime
                prime[j] = False
    return prime


def solve(n, prime):
    res = []
    p = 1
    # Find the largest prime <= 2*n//3
    for i in range(1, 2 * n // 3 + 1):
        if prime[i]:
            p = i
    l, r = p - 1, p + 1
    res.append(p)
    while True:
        if r > n:
            break
        res.append(r)
        r += 1
        if l < 1:
            break
        res.append(l)
        l -= 1
    # Add remaining numbers in order
    res.extend(range(1, l + 1))
    res.extend(range(r, n + 1))
    print(" ".join(map(str, res)))


t = int(input())
prime = soe(limit)
for _ in range(t):
    n = int(input())
    solve(n, prime)


# https://github.com/VaHiX/codeForces/